### DISCUSSION TITLE
Database Hygiene and Zombie Entries

### DISCUSSION THREAD

**Discussion by david-morris** [wagtail/wagtail#8393] [Ideas]
I'm a new user of Wagtail and I'm finding a couple problems which I think are related to not having enough SQL constraints, and not documenting internal Wagtail logic.

Our team is working on implementing new features, building Wagtail into an existing Django backend to act headlessly through an existing GraphQL API.  And that's going fine.  The ease of integration is nice.

But sometimes as we work on the project, existing entries just... break. Changing models sets entries in a state where they can neither be deleted or changed from the CMS panel, yet take up a slug. Older images 404 despite being in the SQL tables and the filesystem, and we don't know why.

So it would be nice to have more database constraints --- so that migrations will fail and rollback if they would leave zombie entries --- and documentation about what invariants the system assumes.

Or maybe there's already documentation about this problem?  Maybe a philosophical stance that makes the code easier to understand?

We would really like to avoid having these problems on prod.

**Comment by lb-** [comment]
@david-morris  Here is a grab back of notes/links, others may chime in with some additional ideas but it would be good to get more specifics of the exact problems you are encountering and when they happen.

Some management commands that may help 
* `fixtree` https://docs.wagtail.org/en/stable/reference/management_commands.html#fixtree - could help with some of the page related issues?
* `wagtail_update_image_renditions` https://docs.wagtail.org/en/latest/reference/management_commands.html#wagtail-update-image-renditions - helps with cleaning up / regenerating image renditions (note: coming in Wagtail 3.0, not in the current release)

There is a very old PR that relates to changing page models and encountering errors, would this be something that may help - https://github.com/wagtail/wagtail/pull/4116 (please comment on that PR if so).


Or do you mean we should add DB constraints on `meta` for some models?
 - there were some issues with us using `UniqueConstraint` alongside `condition` - see links below.

* https://github.com/wagtail/wagtail/issues/8089 - PR https://github.com/wagtail/wagtail/pull/8152
```
    class Meta:
        constraints = [
            models.UniqueConstraint(
                fields=["is_default_site"],
                condition=models.Q(is_default_site=True),
                name="unique_in_default_site",
            ),
        ]
```
